File: algorithm_library/python/data-structure/DisjointSparseTable.py
class DisjointSparseTable:
    def __init__(self, arr, op):
        """
        arr: 配列
        op: 演算 (結合律を満たすこと)
        """
        self.n = len(arr)
        self.arr = arr[:]
        self.op = op

        self.max_k = self.n.bit_length()
        self.table = [[0] * self.n for _ in range(self.max_k)]

        for k in range(self.max_k):
            block_size = 1 << (k + 1)  # ブロック全体の長さ
            half = 1 << k  # 半分の長さ

            for start in range(0, self.n, block_size):
                mid = min(start + half, self.n)
                end = min(start + block_size, self.n)

                # 左側 [start, mid) の累積を計算
                if start < mid:
                    cur = arr[mid - 1]
                    self.table[k][mid - 1] = cur
                    for i in range(mid - 2, start - 1, -1):
                        cur = op(arr[i], cur)
                        self.table[k][i] = cur

                # 右側 [mid, end) の累積を計算
                if mid < end:
                    cur = arr[mid]
                    self.table[k][mid] = cur
                    for i in range(mid + 1, end):
                        cur = op(cur, arr[i])
                        self.table[k][i] = cur

    def query(self, l, r):
        assert 0 <= l < r <= self.n
        if r - l == 1:
            return self.arr[l]

        # クエリに答えるために使うレベル k を求める
        k = (l ^ (r - 1)).bit_length() - 1
        return self.op(self.table[k][l], self.table[k][r - 1])

File: algorithm_library/cpp/DisjointSparseTable.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// S: Disjoint Sparse Table で扱うデータ型
// op: 演算関数 (S op(S, S))
template <class S, S (*op)(S, S)>
class DisjointSparseTable {
    int n;
    vector<vector<S>> table;
    vector<int> log_table;

public:
    DisjointSparseTable() {}
    DisjointSparseTable(const vector<S>& v) {
        n = v.size();
        
        // bit_length相当のテーブルサイズ確保
        int b = 1;
        while (b <= n) b <<= 1;
        log_table.resize(b + 1);
        log_table[1] = 0;
        for (int i = 2; i <= b; i++) log_table[i] = log_table[i >> 1] + 1;

        int max_k = (n == 0) ? 0 : log_table[n] + 1;
        // 配列のサイズが2の累乗でない場合のために少し余裕をもたせる計算
        if ((1 << (max_k - 1)) < n) max_k++;

        table.assign(max_k, vector<S>(n));

        for (int k = 0; k < max_k; k++) {
            int block_size = 1 << (k + 1);
            int half = 1 << k;

            for (int start = 0; start < n; start += block_size) {
                int mid = min(start + half, n);
                int end = min(start + block_size, n);

                // 左側 [start, mid)
                if (start < mid) {
                    table[k][mid - 1] = v[mid - 1];
                    for (int i = mid - 2; i >= start; i--) {
                        table[k][i] = op(v[i], table[k][i + 1]);
                    }
                }

                // 右側 [mid, end)
                if (mid < end) {
                    table[k][mid] = v[mid];
                    for (int i = mid + 1; i < end; i++) {
                        table[k][i] = op(table[k][i - 1], v[i]);
                    }
                }
            }
        }
    }

    S query(int l, int r) {
        if (l == r - 1) return table[0][l];
        int k = log_table[l ^ (r - 1)];
        return op(table[k][l], table[k][r - 1]);
    }
};